azkaban-web-server/src/test/java/azkaban/flowtrigger/TriggerInstanceTest.java

```diff
@@ -0,0 +1,269 @@
+    Date expectedStartTime = getDate(2, 2, 2);
+    Date expectedEndTime = getDate(2, 2, 3);
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(2, 2,
+            2), getDate(2, 2, 3)));
+
+    TriggerInstance ti = null;
+    ti = new TriggerInstance("1", null, "1", 1,
+        "test", dependencyInstanceList, -1, null);
+    assertThat(ti.getStartTime()).isEqualTo(expectedStartTime);
+    assertThat(ti.getEndTime()).isEqualTo(expectedEndTime);
+    dependencyInstanceList.clear();
+
+    expectedStartTime = getDate(2, 2, 2);
+    expectedEndTime = getDate(2, 2, 3);
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.RUNNING, CancellationCause.NONE, getDate(2, 2,
+            2), getDate(2, 2, 3)));
+
+    ti = new TriggerInstance("1", null, "1", 1,
+        "test", dependencyInstanceList, -1, null);
+    assertThat(ti.getStartTime()).isEqualTo(expectedStartTime);
+    assertThat(ti.getEndTime()).isNull();
+    dependencyInstanceList.clear();
+
+    expectedStartTime = getDate(2, 2, 2);
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(2, 2,
+            2), null));
+
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(2, 2,
+            2), null));
+
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(2, 2,
+            2), null));
+    ti = new TriggerInstance("1", null, "1", 1,
+        "test", dependencyInstanceList, -1, null);
+    assertThat(ti.getStartTime()).isEqualTo(expectedStartTime);
+    assertThat(ti.getEndTime()).isNull();
+    dependencyInstanceList.clear();
+
+    expectedStartTime = getDate(2, 2, 2);
+    expectedEndTime = getDate(3, 2, 3);
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(2, 2,
+            3), null));
+
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(3, 2,
+            2), null));
+
+    dependencyInstanceList
+        .add(createTestDependencyInstance(Status.SUCCEEDED, CancellationCause.NONE, getDate(2, 2,
+            2), expectedEndTime));
+    ti = new TriggerInstance("1", null, "1", 1,
+        "test", dependencyInstanceList, -1, null);
+    assertThat(ti.getStartTime()).isEqualTo(expectedStartTime);
+    assertThat(ti.getEndTime()).isEqualTo(expectedEndTime);
+    dependencyInstanceList.clear();
+
+    ti = new TriggerInstance("1", null, "1", 1,
+        "test", dependencyInstanceList, -1, null);
+    assertThat(ti.getStartTime()).isNull();
+    assertThat(ti.getEndTime()).isNull();
+    dependencyInstanceList.clear();
+  }
+
+  @Test
+  public void testTriggerInstanceStatus() throws Exception {
```

Comment:
It looks like this is multiple tests wrapped inside a single test method, is that right? If so, it'll be easier to parallelize if broken out into multiple smaller tests. That would also make it easier for us to identify quickly what is broken and why if a test fails.

Reply:
thanks, updated.